File: JavaScript/0328. Odd Even Linked List.js
// Given a singly linked list, group all odd nodes together followed by the even nodes. Please note here we are talking about the node number and not the value in the nodes.
//
// You should try to do it in place. The program should run in O(1) space complexity and O(nodes) time complexity.
//
// Example:
// Given 1->2->3->4->5->NULL,
// return 1->3->5->2->4->NULL.
//
// Note:
// The relative order inside both the even and odd groups should remain as it was in the input.
// The first node is considered odd, the second node even and so on ...

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

// Time O(n)
// Space O(1)
//
//                odd
// Step 0  head -> 1 -> 2 -> 3 -> 4 -> 5 -> null
//                     even
//
//          evenHead -> 2
//
//
//                     odd
// Step 1  head -> 1 -> 3 -> 4 -> 5 -> null
//                          even
//
//          evenHead -> 2 -> 4
//
//
//                          odd
// Step 2  head -> 1 -> 3 -> 5 -> null
//                                even
//
//          evenHead -> 2 -> 4 -> null
//
//
//                          odd
//  Final  head -> 1 -> 3 -> 5 -> 2 -> 4 -> null
//                             evenHead     even
//
function oddEvenList1(head) {
  if (head == null) return null;

  let odd = head;
  const evenHead = head.next;

  while (odd.next && odd.next.next) {
    const even = odd.next;
    odd.next = odd.next.next;
    odd = odd.next;
    even.next = odd.next;
  }

  odd.next = evenHead;
  return head;
}

File: Python/0328. Odd Even Linked List.py
# Given the head of a singly linked list, group all the nodes with odd indices together followed by the nodes with even indices, and return the reordered list.
# The first node is considered odd, and the second node is even, and so on.
# Note that the relative order inside both the even and odd groups should remain as it was in the input.
# You must solve the problem in O(1) extra space complexity and O(n) time complexity.
#
# Example 1:
#
# Input: head = [1,2,3,4,5]
# Output: [1,3,5,2,4]
#
# Example 2:
#
# Input: head = [2,1,3,5,6,4,7]
# Output: [2,3,6,7,1,5,4]
#
# Constraints:
#
# The number of nodes in the linked list is in the range [0, 10^4].
# -10^6 <= Node.val <= 10^6


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


# Notion

# Time O(n)
# Space O(1)
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        odd = head
        even = head.next

        even_head = even
        while even and even.next:
            odd.next = even.next
            odd = odd.next
            even.next = odd.next
            even = even.next
        odd.next = even_head
        return head
